File: src/api/queue.ts
import { invokeWithTimeout } from "../lib/invoke";
import { withErrorHandling } from "../lib/errorHandler";
import type {
  AddToQueueRequest,
  CompleteAllQueueResponse,
  QueueEntry,
  QueueEntryWithTask,
  RemoveFromQueueRequest,
  ReorderQueueRequest,
  UpdateQueueRequest,
} from "../types/queue";

export const queueApi = {
  /**
   * キュー全体を取得（タスク情報含む）
   */
  async getQueue(): Promise<QueueEntryWithTask[]> {
    return await withErrorHandling(
      () => invokeWithTimeout<QueueEntryWithTask[]>("get_task_queue"),
      "キューの取得に失敗しました"
    );
  },

  /**
   * タスクをキューに追加（ステータスがActiveになる）
   */
  async addToQueue(taskId: string): Promise<QueueEntry> {
    const req: AddToQueueRequest = { taskId };
    return await withErrorHandling(
      () => invokeWithTimeout<QueueEntry>("add_task_to_queue", { req }),
      "タスクのキュー追加に失敗しました"
    );
  },

  /**
   * タスクをキューから削除（ステータスを指定）
   */
  async removeFromQueue(
    taskId: string,
    targetStatus: "draft" | "completed"
  ): Promise<void> {
    const req: RemoveFromQueueRequest = { taskId, targetStatus };
    return await withErrorHandling(
      () => invokeWithTimeout<void>("remove_task_from_queue", { req }),
      "タスクのキュー削除に失敗しました"
    );
  },

  /**
   * タスクをdraftに戻す（タスクプールに戻る）
   */
  async returnToDraft(taskId: string): Promise<void> {
    return this.removeFromQueue(taskId, "draft");
  },

  /**
   * タスクを完了としてマーク
   */
  async markAsCompleted(taskId: string): Promise<void> {
    return this.removeFromQueue(taskId, "completed");
  },

  /**
   * キュー全体をクリア
   */
  async clearQueue(): Promise<void> {
    return await withErrorHandling(
      () => invokeWithTimeout<void>("clear_task_queue"),
      "キューのクリアに失敗しました"
    );
  },

  /**
   * キュー内の全タスクを完了状態にする
   */
  async completeAll(): Promise<CompleteAllQueueResponse> {
    return await withErrorHandling(
      () => invokeWithTimeout<CompleteAllQueueResponse>("complete_all_queue"),
      "全タスクの完了に失敗しました"
    );
  },

  /**
   * タスクのキュー内位置を更新
   */
  async updateQueuePosition(
    taskId: string,
    newPosition: number
  ): Promise<QueueEntry> {
    const req: UpdateQueueRequest = { taskId, newPosition };
    return await withErrorHandling(
      () => invokeWithTimeout<QueueEntry>("update_queue_position", { req }),
      "キュー位置の更新に失敗しました"
    );
  },

  /**
   * キュー全体を一括で並び替え
   */
  async reorderQueue(taskIds: string[]): Promise<QueueEntry[]> {
    const req: ReorderQueueRequest = { taskIds };
    return await withErrorHandling(
      () => invokeWithTimeout<QueueEntry[]>("reorder_task_queue", { req }),
      "キューの並び替えに失敗しました"
    );
  },
};

File: src-tauri/src/commands/queue.rs
use tauri::State;

use crate::models::queue::{
    AddToQueueRequest, CompleteAllQueueResponse, QueueEntry, QueueEntryWithTask,
    RemoveFromQueueRequest, ReorderQueueRequest, UpdateQueueRequest,
};
use crate::service::QueueService;
use crate::DbPool;

/// キュー全体を取得
#[tauri::command]
pub fn get_task_queue(pool: State<DbPool>) -> Result<Vec<QueueEntryWithTask>, String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::get_queue(&mut conn).map_err(|e| e.to_string())
}

/// タスクをキューに追加
#[tauri::command]
pub fn add_task_to_queue(
    pool: State<DbPool>,
    req: AddToQueueRequest,
) -> Result<QueueEntry, String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::add_to_queue(&mut conn, req.task_id).map_err(|e| e.to_string())
}

/// タスクをキューから削除
#[tauri::command]
pub fn remove_task_from_queue(
    pool: State<DbPool>,
    req: RemoveFromQueueRequest,
) -> Result<(), String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::remove_from_queue(&mut conn, req.task_id, req.target_status)
        .map_err(|e| e.to_string())
}

/// キュー全体をクリア
#[tauri::command]
pub fn clear_task_queue(pool: State<DbPool>) -> Result<(), String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::clear_queue(&mut conn).map_err(|e| e.to_string())
}

/// キュー内の全タスクを完了状態にする
#[tauri::command]
pub fn complete_all_queue(pool: State<DbPool>) -> Result<CompleteAllQueueResponse, String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    let completed_count = QueueService::complete_all_queue(&mut conn).map_err(|e| e.to_string())?;
    Ok(CompleteAllQueueResponse { completed_count })
}

/// タスクのキュー内位置を更新
#[tauri::command]
pub fn update_queue_position(
    pool: State<DbPool>,
    req: UpdateQueueRequest,
) -> Result<QueueEntry, String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::update_queue_position(&mut conn, req.task_id, req.new_position)
        .map_err(|e| e.to_string())
}

/// キュー全体を一括で並び替え
#[tauri::command]
pub fn reorder_task_queue(
    pool: State<DbPool>,
    req: ReorderQueueRequest,
) -> Result<Vec<QueueEntry>, String> {
    let mut conn = pool.get().map_err(|e| format!("データベース接続エラー: {}", e))?;
    QueueService::reorder_queue(&mut conn, req.task_ids).map_err(|e| e.to_string())
}
